### DISCUSSION TITLE
Why does InvokeError have from_anyhow instead of implementing From<anyhow::Error>?

### DISCUSSION THREAD

**Discussion by CrendKing** [tauri-apps/tauri#4800] [General]
`InvokeError` has [`from_anyhow`](https://github.com/tauri-apps/tauri/blob/079b1cc06e1e5437686bac9049d5ac569c3f42df/core/tauri/src/hooks.rs#L101), which converts an anyhow error to itself. It is useful, so that user can call an internal functions that returns anyhow error from a `command` function. However, it still requires an explicit `map_error(InvokeError::from)` at the end of every fallible command. Why did tauri [decide](https://github.com/tauri-apps/tauri/commit/d6918efdaf386e0b1075d80430da8e807b10541f#diff-e553290818042c28673e30a083a041810c81b41fb6e511535a3a5794de1f5565) to not implement `From<anyhow::Error>` so that the conversion can be implicit?

**Comment by adamhp** [comment]
Currently struggling with this myself! Would love to have the clean syntax from the implicit conversion.

**Comment by FabianLars** [comment]
Not an actual answer to the "why" since I wasn't involved in the linked change, but i always thought that you're not supposed to use `InvokeError` in commands yourself. It looked more like a tauri internal thing which is exposed because it _is_ part of the public api (and needed by those who don't use the command macro).

So from that POV, not having `From` implementations make sense to prevent unexpected/"hidden" behavior. -> That said, not 100% familiar with the macro etc, so i'm not sure if that would even have direct impact but since InvokeError is what the commands actually return in the end i'd imagine so 🤔

**Reply by CrendKing** [reply]
In that case, I'll avoid using `InvokeError` directly. Thanks!

**Comment by JonasKruckenberg** [comment]
This is not by design, but a consequence of Rust's rules around implementing traits for types from 3rd party crates.

If you look [at the code](https://github.com/tauri-apps/tauri/blob/38d0bed8ebb6c5872fc54903051d146d386b721e/core/tauri/src/hooks.rs#L102)

You will notice we are implementing `From<T> for InvokeError` for all types that can be serialized. This makes total sense as the requirements around _commands_ is that all parameters and return types (including errors) must be serializable/deserializable. So far so good.

Now if we were to try to add another impl

```
impl From<anyhow::Error> for InvokeError {
   fn from(value: anyhow::Error) -> Self {
      todo!()
   }
}
```

You will notice the compiler complains. This is because `anyhow::Error` doesn't not currently implement `Serialize` **but it might start to anytime in the future**
This would then break our code, since we would now have two `From` implementations for the same type (the generic `T` would now also match `anyhow::Error`)
As a consequence the compiler wouldn't know anymore which impl to choose and compilation would fail.

This is why we have the rather verbose `from_anyhow` and `from_serde_json` methods.

If you're familiar with Rust RFCs you might have noticed that this sounds and awful lot like [specialization](https://rust-lang.github.io/rfcs/1210-impl-specialization.html) and indeed this would solve our issue, as the compiler would now accept our two `From` impls, simply choosing the more specific one.

**Comment by aminya** [comment]
I have published `anyhow_serde` as a drop-in replacement for `anyhow` with serde support. 
https://github.com/aminya/anyhow_serde

```toml
[dependencies]
anyhow_serde = "1.0"
```

This adds serde implementation for the error to serialize/deserialize the error.

This allows the user to pass around the error types across the communication boundaries in apps such as Tauri without needing to convert the types. That means the `?` operator can be used for Tauri commands, and any code that uses `anyhow:Error` can become a `tauri::command` without any modification.

https://tauri.app/develop/calling-rust/#error-handling

With this feature
```rs
#[tauri::command]
fn my_custom_command() -> Result<(), anyhow_serde::Error> {
  std::fs::File::open("path/that/does/not/exist")?;
  Ok(())
}
```

Without this feature: each function should be migrated away from Anyhow to be usable:
```rs
#[tauri::command]
fn my_custom_command() -> Result<(), String> {
  std::fs::File::open("path/that/does/not/exist").map_err(|e| e.to_string)?;
  Ok(())
}
```

**Reply by FabianLars** [reply]
mind opening a PR for that to our docs? as an easier alternative for the thiserror instructions for when simple errors are enough.

**Reply by alexanderflink** [reply]
@aminya Thank you! Does `anyhow_serde` work for Tauri v2? When trying out your example:
```rust
#[tauri::command]
fn my_custom_command() -> Result<(), anyhow_serde::Error> {
    std::fs::File::open("path/that/does/not/exist")?;
    Ok(())
}

#[cfg_attr(mobile, tauri::mobile_entry_point)]
pub fn run() {
    tauri::Builder::default()
        .plugin(tauri_plugin_opener::init())
        .invoke_handler(tauri::generate_handler![my_custom_command])
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
}
```
I get:
```
the method blocking_kind exists for reference &Result<(), anyhow_serde::Error>, but its trait bounds were not satisfied
the following trait bounds were not satisfied:
anyhow_serde::Error: Into<InvokeError>
which is required by Result<(), anyhow_serde::Error>: tauri::ipc::private::ResultKind
Result<(), anyhow_serde::Error>: IpcResponse
which is required by &Result<(), anyhow_serde::Error>: tauri::ipc::private::ResponseKind (rustc E0599)
hint: Error originated from macro call here
hint: Error originated from macro call here
```

I am new to Rust and Tauri, but it seems like `Into<InvokeError>` is not implemented for `anyhow_serde::Error`?

**Reply by aminya** [reply]
> mind opening a PR for that to our docs? as an easier alternative for the thiserror instructions for when simple errors are enough.

Sure, I'll open a doc update PR.



> @aminya Thank you! Does `anyhow_serde` work for Tauri v2? When trying out your example:
> 

Thanks for the report. The serialize feature was not enabled by default. I just pushed a patch to fix it.

```toml
[dependencies]
anyhow_serde = "1.0.101"
```
